### ISSUE TITLE
Cross-compiling `raw-dylib` doesn't work on Windows

### ISSUE THREAD

**Issue by mati865** [rust-lang/rust#103939]
I tried this code:

```rust
#![feature(raw_dylib)]

fn main() {
    unsafe {
        MessageBoxA(0, b"hello\0".as_ptr(), "world\0".as_ptr(), 0);
    }
}

#[link(name = "user32", kind = "raw-dylib")]
extern "system" {
    fn MessageBoxA(hwnd: usize, lptext: *const u8, lpcaption: *const u8, flags: u32) -> i32;
}
```

I expected to see this happen:
Successful compilation

Instead, this happened:
```
$ PATH=/d/msys64/mingw32/bin:$PATH /c/Users/mateusz/.cargo/bin/rustc test.rs --target i686-pc-windows-gnu
error: Error calling dlltool: program not found

error: aborting due to previous error
```
Worth mentioning it works fine when building natively from x86_64 to x86_64 target.

### Meta

`rustc --version --verbose`:
```
rustc 1.67.0-nightly (edf018221 2022-11-02)
binary: rustc
commit-hash: edf0182213a9e30982eb34f3925ddc4cf5ed3471
commit-date: 2022-11-02
host: x86_64-pc-windows-gnu
release: 1.67.0-nightly
LLVM version: 15.0.4
```

**Comment by bjorn3** [comment]
https://github.com/rust-lang/rust/blob/31d754a1dfba4d091a05589459708f286d5a231e/compiler/rustc_codegen_llvm/src/back/archive.rs#L406

Looks like rustc expects a toolchain that cross compiles from the host (x86_64 mingw) to the target (i686 mingw) rather than a toolchain native to the target (i686 mingw) like in your case. I guess it could try both executable names.

**Comment by mati865** [comment]
On Windows toolchains commonly provide only `dlltool` without prefix but it's possible to cross-generate import library using dlltool for any host with trick discovered by glandium: [microsoft/windows-rs@`cb8d6a9` (#2016)](https://github.com/microsoft/windows-rs/pull/2016/commits/cb8d6a9181a50903589bfecf962b464a45f98fcf#diff-82947385e42042ffba4e2ba53c683427146ce3bf9e7c15ef45ad38e624a66444R94-R106)

**Comment by mati865** [comment]
I have it on my radar but cannot tell when when I can find free time to do it.

**Comment by eiderdaus** [comment]
I am unsure whether the linked pull request above actually fixed this issue for cross-compilation, but I get the same exact error with the stable-x86_64-pc-windows-gnu toolchain version 1.71.0. In this version, the feature should have been stabilized. Should I open a new issue because this one only regards cross-compilation?

**Comment by mati865** [comment]
Looking at https://github.com/tokio-rs/mio/issues/1632#issuecomment-1635632791 your issue is a bit different. How did you install mingw-w64 toolchain?
MSYS2 and mingw-w64 builds provide `dlltool.exe` but other toolchains might provide `x86_64-w64-mingw32-dlltool.exe` instead. In such case the easiest workaround is to symlink `dlltool.exe` pointing to your `*-dlltool.exe`.

**Comment by mati865** [comment]
Alternatively you could try `RUSTFLAGS='-Cdlltool=<your-dlltool-bin>'`.

**Comment by eiderdaus** [comment]
Hi, thanks for the prompt responses! My host is Windows, there's no dlltool on my %PATH%. I used Rustup to install Rust and I found a dlltool.exe in the following folder:

%USERPROFILE%\.rustup\toolchains\stable-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\bin\self-contained

I would expect that this should be enough for building working Rust applications. I tried passing this dlltool via `RUSTFLAGS` and got an error

    error: Dlltool could not create import library:
           [redacted]\.rustup\toolchains\stable-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\bin\self-contained\dlltool.exe: CreateProcess

> How did you install mingw-w64 toolchain?

I never mentioned any mingw-w64 but actually I do have MSYS2 at my disposal. It's not on my %PATH% at the moment. I just used `RUSTFLAGS=-Cdlltool=C:\msys64\clang64\bin\dlltool.exe` with success - however I would call this a workaround, not a solution!

**Comment by dpaoliello** [comment]
> I never mentioned any mingw-w64 but actually I do have MSYS2 at my disposal. It's not on my %PATH% at the moment. I just used `RUSTFLAGS=-Cdlltool=C:\msys64\clang64\bin\dlltool.exe` with success - however I would call this a workaround, not a solution!

Correct, the use of `dlltool` is considered to be a temporary workaround until the minimum required binutils version supports the import libraries that LLVM generates: https://github.com/rust-lang/rust/pull/104218#issuecomment-1315895618

I the meantime, having dlltool on your `PATH`, or using `-Cdlltool` is required.

**Comment by mati865** [comment]
> I would expect that this should be enough for building working Rust applications.

I'd expect that as well.



> I tried passing this dlltool via `RUSTFLAGS` and got an error
> 
> ```
> error: Dlltool could not create import library:
>        [redacted]\.rustup\toolchains\stable-x86_64-pc-windows-gnu\lib\rustlib\x86_64-pc-windows-gnu\bin\self-contained\dlltool.exe: CreateProcess
> ```

This definitely doesn't seem right, might be worth to create issue for having `self-contained` dlltool working by the default.

**Comment by eiderdaus** [comment]
> I the meantime, having dlltool on your `PATH`, or using `-Cdlltool` is required.

Alright, thanks a lot for this information. To be honest I am a little bit disenchanted after the great feature announcement on the Rust blog - I don't normally expect a stabilized feature to still require such workarounds.

> This definitely doesn't seem right, might be worth to create issue for having `self-contained` dlltool working by the default.

Not sure if another issue makes much sense when the whole dlltool invocation should be removed either way if I understood correctly?

**Comment by mati865** [comment]
> Not sure if another issue makes much sense when the whole dlltool invocation should be removed either way if I understood correctly?

Removal of dlltool might take years, in the meantime bundled dlltool should be fixed and used by the default.

**Comment by eiderdaus** [comment]
> Removal of dlltool might take years, in the meantime bundled dlltool should be fixed and used by the default.

I see. Just so I fully understand: If what I described is a new issue in your opinion, then why is this one still open?

**Comment by ChrisDenton** [comment]
It's been a year since this issue was originally reported. Can we bump the self-contained binutils at least?

**Comment by mati865** [comment]
> Can we bump the self-contained binutils at least?

That's my plan for the near future but it's easier said than done. Though, at current pace it should hit nightly before 2024.

**Comment by ChrisDenton** [comment]
Thanks for the quick reply! And sorry that updating is still an ordeal.

**Comment by kleisauke** [comment]
Just wondering, perhaps we could use https://github.com/messense/implib-rs to generate a GNU binutils flavored import library? It seems to work well with [cargo-c](https://github.com/lu-zero/cargo-c) for the `x86_64-pc-windows-gnu` target, although the `i686-pc-windows-gnu` target required a [minor patch](https://github.com/messense/implib-rs/pull/13) to function properly.

**Comment by mati865** [comment]
Does it output import libraries in legacy format in GNU mode?
Naming the modes GNU and MSVC is confusing because GCC didn't invent its own format. It just implemented what MSVC was using two decades ago. Currently, MSVC uses a short import library format that other tools like LLVM mimic.

Ideally, Rust could say that it only works with Binutils that properly support short import libraries, that is 2.40 released almost exactly one year ago. That would free it from all this dlltool nonsense and just rely on the backend like other Windows targets do.

**Comment by ChrisDenton** [comment]
We'd probably need an MCP to document a minimum supported binutils version but personally I'd be very much in favour of that.

**Comment by kleisauke** [comment]
> Does it output import libraries in legacy format in GNU mode?

Yes, it does. LLVM (and likely MSVC) can link to this without any issues. See for example the simplified testcase here:
https://gist.github.com/kleisauke/38f4a33b77e9f09669c23c8a35f64752

However, the `implib` crate does not support Arm64EC/Arm64X. So, bumping the minimum supported version of Binutils in Rust to 2.40 is probably the best solution.

**Comment by mati865** [comment]
> Yes, it does. LLVM (and likely MSVC) can link to this without any issues. See for example the simplified testcase here:
> [gist.github.com/kleisauke/38f4a33b77e9f09669c23c8a35f64752](https://gist.github.com/kleisauke/38f4a33b77e9f09669c23c8a35f64752)

MSVC and LLVM can use both legacy and short import libs without a problem, I'm worried only about ld.bfd. Some ld.bfd bugs happen only in non-obvious scenarios, like when an import library is included within another library (fixed in 2,40 by Martin).

> However, the `implib` crate does not support Arm64EC/Arm64X. So, bumping the minimum supported version of Binutils in Rust to 2.40 is probably the best solution.

That wouldn't be a problem if `implib` was used only on `windows-gnu`.

**Comment by workingjubilee** [comment]
Note that per https://github.com/rust-lang/rust/issues/140704 it doesn't seem to require cross-compilation to trigger this anymore.

**Comment by mati865** [comment]
Nothing has changed here in Rust. The only difference is the C toolchain: there is no consensus whether dlltool should be called `dlltool.exe` or `<triple>-dlltool.exe` and toolchain providers typically provide just one variant of dlltool name.

Using the same C toolchain as in https://github.com/rust-lang/rust/issues/140704, back in 2022 would hit the same issue.

**Comment by mati865** [comment]
Yeah, that would be a problem. I guess nobody expected the dlltool to be provided by a binary that consists of multiple tools.

**Comment by bjorn3** [comment]
You can use `CARGO_ENCODED_RUSTFLAGS` instead if you need a rustc flag containing spaces. This uses a specific ascii control character as argument separator instead of space. You can find which in the cargo docs.

**Comment by kleisauke** [comment]
> However, the `implib` crate does not support Arm64EC/Arm64X.

This is no longer an issue with [`implib` v0.5.0](https://github.com/messense/implib-rs/releases/tag/v0.5.0) (which also uses [`ar_archive_writer`](https://github.com/rust-lang/ar_archive_writer) since [v0.6.0](https://github.com/messense/implib-rs/releases/tag/v0.6.0)).
